File: app/src/main/java/net/dixq/unlimiteddiary/authentication/AuthenticateActivity.java
package net.dixq.unlimiteddiary.authentication;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.os.Bundle;
import android.os.FileUtils;
import android.os.Handler;

import androidx.appcompat.app.AppCompatActivity;

import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.services.drive.Drive;
import com.google.api.services.drive.DriveScopes;

import net.dixq.unlimiteddiary.google_api.DriveHelper;
import net.dixq.unlimiteddiary.singleton.ApiAccessor;
import net.dixq.unlimiteddiary.top.TopActivity;
import java.util.Arrays;

public class AuthenticateActivity extends AppCompatActivity {

    public static final int REQUEST_CODE_TOP = 0;
    public static final int REQUEST_ACCOUNT_PICKER = 1;
    private GoogleAccountCredential _credential;
    private Handler _handler = new Handler();

    @Override
    protected void onCreate(Bundle b){
        super.onCreate(b);
        setupAuthentication();
    }

    private void setupAuthentication() {
        _credential = GoogleAccountCredential.usingOAuth2(this, Arrays.asList(DriveScopes.DRIVE,"https://www.googleapis.com/auth/photoslibrary"));
        String accountName = getSavedAccountName();
        if (accountName.isEmpty()) {
            startActivityForResult(_credential.newChooseAccountIntent(), AuthenticateActivity.REQUEST_ACCOUNT_PICKER);
        } else {
            _credential.setSelectedAccountName(accountName);
            proceesNextStep();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case AuthenticateActivity.REQUEST_ACCOUNT_PICKER:
                if (resultCode == RESULT_OK && data != null) {
                    String accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
                    String accountType = data.getStringExtra(AccountManager.KEY_ACCOUNT_TYPE);
                    saveAccountName(accountName);
                    saveAccountType(accountType);
                    _credential.setSelectedAccountName(accountName);
                    proceesNextStep();
                } else {
                    finish();
                }

                break;
            case REQUEST_CODE_TOP:
                finish();
                break;
        }
    }

    private void proceesNextStep(){
        ApiAccessor.getInstance().setDriveService(getDriveService());
        new Thread(()->{
            DriveHelper driveHelper = new DriveHelper(ApiAccessor.getInstance());
            ApiAccessor.getInstance().setFolderId(driveHelper.getFolderId());
            _handler.post(this::startNextActivity);
        }).start();
    }

    private void startNextActivity(){
        Intent intent = new Intent(this, TopActivity.class);
        startActivityForResult(intent, REQUEST_CODE_TOP);
    }

    private Drive getDriveService() {
        return new Drive.Builder(AndroidHttp.newCompatibleTransport(), new GsonFactory(), _credential).build();
    }

    private static final String TAG_AUTH_ACOUNT = "TAG_AUTH_ACOUNT";
    private static final String KEY_AUTH_ACOUNT_NAME = "KEY_AUTH_ACOUNT_NAME";
    private static final String KEY_AUTH_ACOUNT_TYPE = "KEY_AUTH_ACOUNT_TYPE";
    private String getSavedAccountName() {
        return getSharedPreferences(TAG_AUTH_ACOUNT, Context.MODE_PRIVATE)
                .getString(KEY_AUTH_ACOUNT_NAME, "");
    }
    private void saveAccountName(final String accountName) {
        getSharedPreferences(TAG_AUTH_ACOUNT, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY_AUTH_ACOUNT_NAME, accountName)
                .apply();
    }
    private String getSavedAccountType() {
        return getSharedPreferences(TAG_AUTH_ACOUNT, Context.MODE_PRIVATE)
                .getString(KEY_AUTH_ACOUNT_TYPE, "");
    }
    private void saveAccountType(final String accountType) {
        getSharedPreferences(TAG_AUTH_ACOUNT, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY_AUTH_ACOUNT_TYPE, accountType)
                .apply();
    }

}

File: app/src/main/java/net/dixq/unlimiteddiary/authentication/AuthenticateActivity.kt
package net.dixq.unlimiteddiary.authentication

import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.Scope
import com.google.android.material.textfield.TextInputEditText
import com.google.api.client.extensions.android.http.AndroidHttp
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.json.gson.GsonFactory
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import net.dixq.unlimiteddiary.R
import net.dixq.unlimiteddiary.common.Lg
import net.dixq.unlimiteddiary.common.OkDialog
import net.dixq.unlimiteddiary.common.PrefUtils
import net.dixq.unlimiteddiary.common.google_api.DriveHelper
import net.dixq.unlimiteddiary.common.singleton.ApiAccessor
import net.dixq.unlimiteddiary.top.TopActivity
import java.util.*


class AuthenticateActivity : AppCompatActivity(), View.OnClickListener {

    private var _credential: GoogleAccountCredential? = null
    private val _handler = Handler()
    private var _isAuthenticationReady = false
    private var _isNameReady = false
    private var _buttons = Buttons()

    override fun onCreate(b: Bundle?) {
        super.onCreate(b)
        if(!isGoogleApiAvailable()){
            OkDialog(this, "GooglePlay開発者サービスをインストールするか有効にしてください。",  DialogInterface.OnClickListener { _, _ ->
                startGooglePlay()
                finish()
            }).show()
            return
        }
        if (PrefUtils.read(this, KEY_HANDLE_NAME).isEmpty()) {
            setContentView(R.layout.main_authentication)
            findViewById<View>(R.id.button_submit).setOnClickListener(this)
            _buttons.layout(this)
        } else {
            _isNameReady = true
        }
        setupAuthentication()
    }

    private fun isGoogleApiAvailable() :Boolean{
        val availability = GoogleApiAvailability.getInstance()
        return when(availability.isGooglePlayServicesAvailable(this)){
            ConnectionResult.SUCCESS -> true
            ConnectionResult.SERVICE_DISABLED,
            ConnectionResult.SERVICE_MISSING,
            ConnectionResult.SERVICE_UPDATING,
            ConnectionResult.SERVICE_VERSION_UPDATE_REQUIRED,
            ConnectionResult.SERVICE_INVALID,
            ConnectionResult.SERVICE_DISABLED -> false
            else -> false
        }
    }

    private fun startGooglePlay() {
        val googlePlayIntent = Intent(Intent.ACTION_VIEW);
        googlePlayIntent.data = Uri.parse("market://details?id=com.google.android.gms")
        startActivityForResult(googlePlayIntent, REQUEST_GOOGLE_PLAY);
    }

    private fun setupAuthentication() {
        val account: GoogleSignInAccount? = GoogleSignIn.getLastSignedInAccount(this)
        if(account==null) {
            val googleSignInOptions = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestScopes(Scope(DriveScopes.DRIVE))
                .build()
            var googleSignInClient = GoogleSignIn.getClient(this, googleSignInOptions)
            startActivityForResult(googleSignInClient.signInIntent, REQUEST_ACCOUNT_PICKER)
        } else {
            // ログインしていれば、認証スキップ
            _credential = GoogleAccountCredential.usingOAuth2(
                this, Collections.singleton(DriveScopes.DRIVE)
            )
            _credential!!.selectedAccount = account.account
            _isAuthenticationReady = true
            proceesNextStep()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_ACCOUNT_PICKER-> {
                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
                try {
                    val account: GoogleSignInAccount? = task.getResult(ApiException::class.java)
                    Lg.e("成功")
                    _credential = GoogleAccountCredential.usingOAuth2(
                        this, Collections.singleton(DriveScopes.DRIVE)
                    )
                    _credential!!.selectedAccount = account!!.account
                    _isAuthenticationReady = true
                    proceesNextStep()
                } catch (e: Exception) {
                    Lg.e("失敗")
                }
            }
            else -> {
                Lg.e("その他！");
            }
        }
    }

    override fun onClick(v: View) {
        val edt = findViewById<TextInputEditText>(R.id.edt_body)
        if (edt.text.toString().isEmpty()) {
            OkDialog(
                this,
                "ニックネームを入力してください。",
                null
            ).show()
            return
        }
        val color = _buttons.getColor()
        if(color == -1){
            OkDialog(this, "カラーを選択してください。", null).show()
            return
        }
        PrefUtils.save(this, KEY_HANDLE_NAME, edt.text.toString())
        PrefUtils.save(this, KEY_HANDLE_NAME_COLOR, color.toString())
        _isNameReady = true
        proceesNextStep()
    }

    private fun proceesNextStep() {
        if (!_isAuthenticationReady || !_isNameReady) {
            return
        }
        ApiAccessor.getInstance().driveService = driveService
        Thread(Runnable {
            val driveHelper = DriveHelper(ApiAccessor.getInstance())
            ApiAccessor.getInstance().folderId = driveHelper.getFolderId()
            _handler.post { startNextActivity() }
        }).start()
    }

    private fun startNextActivity() {
        val intent = Intent(this, TopActivity::class.java)
        startActivityForResult(intent, REQUEST_CODE_TOP)
    }

    private val driveService: Drive
        private get() = Drive.Builder(
            AndroidHttp.newCompatibleTransport(),
            GsonFactory(),
            _credential
        ).build()

    companion object {
        const val REQUEST_CODE_TOP = 0
        const val REQUEST_ACCOUNT_PICKER = 1
        const val REQUEST_GOOGLE_PLAY = 2
        public const val KEY_HANDLE_NAME = "KEY_HANDLE_NAME"
        public const val KEY_HANDLE_NAME_COLOR = "KEY_HANDLE_NAME_COLOR"
    }
}